utils/s2n_mem.c

```diff
@@ -27,21 +27,87 @@
 static long page_size = 4096;
 static int use_mlock = 1;
 
-int s2n_mem_init(void)
+static int s2n_mem_init_impl(void)
 {
     GUARD(page_size = sysconf(_SC_PAGESIZE));
     if (getenv("S2N_DONT_MLOCK")) {
         use_mlock = 0;
     }
-
-    return 0;
+    return S2N_SUCCESS;
 }
 
-int s2n_mem_cleanup(void)
+static int s2n_mem_cleanup_impl(void)
 {
     page_size = 4096;
     use_mlock = 1;
-    return 0;
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_free_impl(void *ptr, uint32_t size)
+{
+    int munlock_rc = use_mlock ? munlock(ptr, size) : 0;
+    free(ptr);
+    GUARD(munlock_rc);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_malloc_impl(void **ptr, uint32_t requested, uint32_t *allocated)
+{
+    notnull_check(ptr);
+
+    if(!use_mlock) {
```

Comment:
Proposal: instead of doing this as a conditional in the code, have two implementations, one of which mlocks, and one which doesn't.  Register the chosen implementation at startup